certbot-dns-cloudflare/certbot_dns_cloudflare/_internal/dns_cloudflare.py

```diff
@@ -170,22 +177,23 @@ def _find_zone_id(self, domain):
             try:
                 zones = self.cf.zones.get(params=params)  # zones | pylint: disable=no-member
             except CloudFlare.exceptions.CloudFlareAPIError as e:
-                code = int(e)
-                hint = None
-
-                if code == 6003:
-                    hint = ('Did you copy your entire API token/key? To use Cloudflare tokens, '
-                            'you\'ll need python-cloudflare>=2.3.1. You have python-cloudflare {}'
-                            .format(CloudFlare.__version__))
-                elif code == 9103:
-                    hint = 'Did you enter the correct email address?'
-                elif code == 9109:
-                    hint = ('Does your API token have "Zone:Zone:Read" and "Zone:DNS:Edit" '
-                            'permissions for all zones?')
-
-                raise errors.PluginError('Error determining zone_id: {0} {1}. Please confirm that '
-                                         'you have supplied valid Cloudflare API credentials.{2}'
-                                         .format(code, e, ' ({0})'.format(hint) if hint else ''))
+                if 'com.cloudflare.api.account.zone.list' not in str(e):
+                    code = int(e)
+                    hint = None
+
+                    if code == 6003:
+                        hint = ('Did you copy your entire API token/key? To use Cloudflare tokens, '
+                        'you\'ll need the python package cloudflare>=2.3.1.{}'
+                        .format(' This certbot is running cloudflare ' + str(CloudFlare.__version__)
+                        if hasattr(CloudFlare, '__version__') else ''))
+                    elif code == 9103:
+                        hint = 'Did you enter the correct email address and Global key?'
+                    elif code == 9109:
+                        hint = 'Did you enter a valid Cloudflare Token?'
+
+                    raise errors.PluginError('Error determining zone_id: {0} {1}. Please confirm '
+                                     'that you have supplied valid Cloudflare API credentials.{2}'
+                                          .format(code, e, ' ({0})'.format(hint) if hint else ''))
 
             if zones:
```

Comment:
I think we should set `zones` to an empty list before this `for` loop to match the case where `zones.get` returns an empty list when using an email address and API key. When testing this with an API token triggering the permissions error about `com.cloudflare.api.account.zone.list`, Certbot crashed with:

`UnboundLocalError: local variable 'zones' referenced before assignment`